---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Visualization of the Genotype Matrix
```{r, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F, eval = T, results = 'hide',
                      fig.align = 'center')
knitr::opts_knit$set(root.dir = here::here())
```

```{r}
library(tidyverse)
library(vcfR)
library(Biostrings)
source("R/vcf_utils.R")
```

```{r}
#.................................
# Read in Metadata
#.................................
mtdt <- readRDS("data/derived_data/mtdt.RDS")
```




## Haplotype Visualization

### Raw(ish) Calls
I have subsetted to SNPs and excluded samples with more than 20% heterozygous sites (i.e. I think there is MOI or more technically, heteroplasmy). 
```{r}
#........................................................
# Read in long vcf
#........................................................
vcf.snp.ss.long_gt <- readRDS("data/derived_data/final_vcf_snps_long.RDS")
vcf.snp.ss.long_gt <- vcf.snp.ss.long_gt %>% 
  dplyr::left_join(., y = mtdt, by = c("iid", "country", "vividregion")) %>% 
  dplyr::filter(passed == "Y") %>% 
  dplyr::mutate(POS = factor(POS, levels = c(1:5989), ordered = T))  

gtplot <- vcf.snp.ss.long_gt %>% 
  dplyr::mutate(country = ifelse(vividregion == "Lab", iid, country),
                country = stringr::str_replace_all(country, "_", " ")) %>% 
  ggplot() + 
  geom_tile(aes(x = smpls, y = POS, fill = gt_GT)) + 
  scale_fill_brewer("GT Calls", type = "qual", palette = "Dark2") +
  facet_wrap(~vividregion + country, scales = "free_x", nrow = 1, strip.position = "top") +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank())


jpeg("figures/raw_all_samples_gtplot.jpg", height = 8, width = 14, units = "in", res = 500)
plot(gtplot)
graphics.off()



```


```{r, results='asis', fig.width=11, fig.height=8, fig.align='center'}
plot(gtplot)
```


### Filtered Calls, All Natural/Globally Sourced Samples
```{r}
mtdna <- Biostrings::readDNAStringSet(filepath = "data/fasta/mtdna_anc.fa")
# going to drop cynomolgi and lab strains here
mtdna <- mtdna[!names(mtdna) %in% c("SRS417748",
                                    "SRS365051",
                                    "SRS365050",
                                    "SRS258178",
                                    "SRS258349",
                                    "SRS258091",
                                    "SRS693469",
                                    "SRS693266",
                                    "SRS693348",
                                    "SRS693899",
                                    "SRS258286",
                                    "ERS001838")]

seqmat <-  tibble::as_tibble( t(as.matrix(mtdna) ) )
seqmat <- cbind.data.frame(POS = 1:nrow(seqmat), seqmat)

segsites <- apply(seqmat[,2:ncol(seqmat)], 1, function(x){ return( length(unique(x)) > 1 ) })

seqmat.long <- seqmat %>% 
  dplyr::filter(segsites) %>% 
  tidyr::gather(., key = "smpls", value = "GT", 2:ncol(.)) %>% 
  dplyr::left_join(., mtdt, by = "smpls") %>% 
  dplyr::filter(passed == "Y") %>% 
  dplyr::mutate(POS = factor(POS, levels = c(1:5989), ordered = T)) 

gtplot <- seqmat.long %>% 
  dplyr::mutate(GT = ifelse(GT == "N", NA, GT)) %>% 
  ggplot() + 
  geom_tile(aes(x = smpls, y = POS, fill = GT)) + 
  scale_fill_manual("Alleles", values = c("#4daf4a", "#377eb8", "#ffff33", "#e41a1c")) +
  facet_wrap(~vividregion + country, scales = "free_x", nrow = 1, strip.position = "top") +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank())

jpeg("figures/filtered_all_samples_gtplot.jpg", height = 8, width = 14, units = "in", res = 500)
plot(gtplot)
graphics.off()

```
```{r, results='asis', fig.width=11, fig.height=8, fig.align='center'}
plot(gtplot)
```

### Segsites, All Globally Sourced Samples
```{r}

seqmat <-  tibble::as_tibble( t(as.matrix(mtdna) ) )
seqmat <- cbind.data.frame(POS = 1:nrow(seqmat), seqmat)

seqmat <- seqmat[, !colnames(seqmat) %in% mtdt$smpls[mtdt$vividregion == "Lab"]]

segsites <- apply(seqmat[,2:ncol(seqmat)], 1, function(x){ return( length(unique(x)) > 1 ) })


seqmat.long <- seqmat %>% 
  dplyr::filter(segsites) %>% 
  tidyr::gather(., key = "smpls", value = "GT", 2:ncol(.)) %>% 
  dplyr::left_join(., mtdt, by = "smpls") %>% 
  dplyr::filter(passed == "Y") %>% 
  dplyr::mutate(POS = factor(POS, levels = c(1:5989), ordered = T)) 

gtplot <- seqmat.long %>% 
  dplyr::mutate(GT = ifelse(GT == "N", NA, GT)) %>% 
  ggplot() + 
  geom_tile(aes(x = smpls, y = POS, fill = GT)) + 
  scale_fill_manual("Alleles", values = c("#4daf4a", "#377eb8", "#ffff33", "#e41a1c")) +
  facet_wrap(~k + country, scales = "free_x", nrow = 1, strip.position = "top") +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank())

jpeg("figures/filtered_segsites_samples_gtplot.jpg", height = 8, width = 14, units = "in", res = 500)
plot(gtplot)
graphics.off()

```
```{r, results='asis', fig.width=11, fig.height=8, fig.align='center'}
plot(gtplot)

```



### Filtered Calls, Removed Clones 
```{r}

seqmat <-  tibble::as_tibble( t(as.matrix(mtdna) ) )
seqmat <- cbind.data.frame(POS = 1:nrow(seqmat), seqmat)

segsites <- apply(seqmat[,2:ncol(seqmat)], 1, function(x){ return( length(unique(x)) > 1 ) })


seqmat.long <- seqmat %>% 
  dplyr::filter(segsites) %>% 
  tidyr::gather(., key = "smpls", value = "GT", 2:ncol(.)) %>% 
  dplyr::mutate(GT = ifelse(GT == "N", NA, GT)) %>% 
  dplyr::mutate(GT = factor(GT)) %>% 
  dplyr::left_join(., mtdt, by = "smpls") %>% 
  dplyr::filter(hapuid == "Y") %>% 
  dplyr::mutate(POS = factor(POS, levels = c(1:5989), ordered = T)) 

gtplot <- seqmat.long %>% 
  ggplot() + 
  geom_tile(aes(x = smpls, y = POS, fill = GT)) + 
  scale_fill_manual("Alleles", values = c("#4daf4a", "#377eb8", "#ffff33", "#e41a1c")) +
  facet_wrap(~k+country, scales = "free_x", nrow = 1, strip.position = "top") +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank())

jpeg("figures/filtered_noclones_samples_gtplot.jpg", height = 8, width = 14, units = "in", res = 500)
plot(gtplot)
graphics.off()

```

```{r, results='asis', fig.width=11, fig.height=8, fig.align='center'}

plot(gtplot)

```



### Filtered Calls, Removed Clones and NHA
```{r}

seqmat <-  tibble::as_tibble( t(as.matrix(mtdna) ) )
seqmat <- cbind.data.frame(POS = 1:nrow(seqmat), seqmat)

seqmat <- seqmat[, !colnames(seqmat) %in% mtdt$smpls[mtdt$vividregion == "NHA"]]

segsites <- apply(seqmat[,2:ncol(seqmat)], 1, function(x){ return( length(unique(x)) > 1 ) })


seqmat.long <- seqmat %>% 
  dplyr::filter(segsites) %>% 
  tidyr::gather(., key = "smpls", value = "GT", 2:ncol(.)) %>% 
  dplyr::mutate(GT = ifelse(GT == "N", NA, GT)) %>% 
  dplyr::mutate(GT = factor(GT)) %>% 
  dplyr::left_join(., mtdt, by = "smpls") %>% 
  dplyr::filter(hapuid == "Y") %>% 
  dplyr::mutate(POS = factor(POS, levels = c(1:5989), ordered = T)) 

gtplot <- seqmat.long %>% 
  ggplot() + 
  geom_tile(aes(x = smpls, y = POS, fill = GT)) + 
  scale_fill_manual("Alleles", values = c("#4daf4a", "#377eb8", "#ffff33", "#e41a1c")) +
  facet_wrap(~k+country, scales = "free_x", nrow = 1, strip.position = "top") +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank())

jpeg("figures/filtered_noclones_nonha_samples_gtplot.jpg", height = 8, width = 14, units = "in", res = 500)
plot(gtplot)
graphics.off()



```

```{r, results='asis', fig.width=11, fig.height=8, fig.align='center'}

plot(gtplot)

```
