---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Phylogenetic Analyses, All Filtered Samples
```{r, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F, eval = T, results = 'hide',
                      fig.align = 'center', fig.width = 8, fig.height = 8)
```


```{r}
library(tidyverse)
library(Biostrings)
library(ape)
library(ggtree)
set.seed(48)
```

```{r}
#---------------------------------------------------- 
# Read In
#---------------------------------------------------- 
mtdna <- Biostrings::readDNAStringSet(filepath = "~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/data/fasta/mtdna_anc.fa")

# going to drop the ancestral haplo here because I will do an unrooted tree
# since we are not estimating the molecular clock
mtdna <- mtdna[!names(mtdna) %in% "Anc"]

smpls <- readxl::read_excel("~/Documents/GitHub/VivID_Seq/scrape_pubseqs/vivid_seq_public_NGS.xlsx") %>% 
  dplyr::rename(smpls = acc) %>% 
  dplyr::select(c("smpls","country", "vividregion"))

fnlsmpls <- readRDS("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/data/derived_data/final_smpl_list.RDS")

fnlsmpls <- data.frame(smpls = fnlsmpls, stringsAsFactors = F) 
fnlsmpls <- dplyr::left_join(fnlsmpls, smpls, by = "smpls")
fnlsmpls$vividregion[is.na(fnlsmpls$vividregion)] <- "DRC" 
fnlsmpls$country[fnlsmpls$vividregion == "DRC"] <- "CD"

```

### Haplotype Viz
```{r}

seqmat <-  tibble::as_tibble( t(as.matrix(mtdna) ) )
seqmat <- cbind.data.frame(POS = 1:nrow(seqmat), seqmat)

seqmat <- seqmat[, !colnames(seqmat) %in% fnlsmpls$smpls[fnlsmpls$country == "GA"]]

segsites <- apply(seqmat[,2:ncol(seqmat)], 1, function(x){ return( length(unique(x)) > 1 ) })



seqmat.long <- seqmat %>% 
  dplyr::filter(segsites) %>% 
  tidyr::gather(., key = "smpls", value = "GT", 2:ncol(.)) %>% 
  dplyr::mutate(GT = factor(GT)) %>% 
  dplyr::left_join(., fnlsmpls, by = "smpls") %>% 
  dplyr::mutate(POS = factor(POS, levels = c(1:5989), ordered = T)) 

gtplot <- seqmat.long %>% 
  ggplot() + 
  geom_tile(aes(x = smpls, y = factor(POS), fill = GT)) + 
  scale_fill_manual(values = c("#4daf4a", "#377eb8", "#ffff33", "#d9d9d9", "#e41a1c")) +
  facet_wrap(~country, scales = "free_x", nrow = 1, strip.position = "top") +
  theme(axis.text = element_blank(),
        axis.title.x = element_blank())

dir.create("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/")
jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/gtplot.jpg", height = 8, width = 11, units = "in", res = 500)
plot(gtplot)
graphics.off()



```


### Distance Metrics
```{r}
mtdna.unique.ape <- ape::as.DNAbin(mtdna)

mtdna.dist.JC <- ape::dist.dna(mtdna.unique.ape, model="JC69")
mtdna.dist.n <- ape::dist.dna(mtdna.unique.ape, model="N")
```


```{r, results='asis', fig.width=11, fig.height=8}

# plot
JC69plot <- broom::tidy(mtdna.dist.JC) %>% 
  magrittr::set_colnames(c("smpl1", "smpl2", "dist")) %>% 
  ggplot(data=.) +
  geom_tile(aes(x=smpl1, y=smpl2, fill=dist)) +
  ggtitle("Genetic Distance under Jukes-Cantor's 1969 Model") +
  scale_fill_gradient(low = "#fee0d2", high = "#de2d26") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 5),
        axis.text.x = element_text(angle=90, size = 5),
        axis.title = element_blank(),
        plot.title = element_text(family = "Arial", face = "bold", size = 14, hjust=0.5))
  
pairwiseplot <- broom::tidy(mtdna.dist.n) %>% 
  magrittr::set_colnames(c("smpl1", "smpl2", "dist")) %>% 
  ggplot(data=.) +
  geom_tile(aes(x=smpl1, y=smpl2, fill=dist)) +
  ggtitle("Genetic Pairwise Distance \n (Hammings)") +
  scale_fill_gradient(low = "#fee0d2", high = "#de2d26") +
  theme_minimal() +
  theme(axis.text.y = element_text(size = 5),
        axis.text.x = element_text(angle=90, size = 5),
        axis.title = element_blank(),
        plot.title = element_text(family = "Arial", face = "bold", size = 14, hjust=0.5))
  
gridExtra::grid.arrange(JC69plot, pairwiseplot, ncol=2)


```


### Analyze Dist Model Fit

Here we will check the tree versus distance correlation matrix to determine if the Jukes-Cantor approach is appropriate as was done in [Rodrigues et al. 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5792595/). Follow Thibaut's [tutorial](http://adegenet.r-forge.r-project.org/files/MRC-session2-tuto.1.3.pdf) for this (Thanks, Thibaut)!
Generate a NJ tree (purpose isn't to be informative), but need topology for distance metric fit correlation check.


```{r, results='asis'}

par(mfrow=c(1,2))

tree <- ape::njs(mtdna.dist.JC)

x <- as.vector(mtdna.dist.JC) # collapse this distance matrix
y <- as.vector(as.dist(cophenetic(tree))) # collapse tree distance matrix, cophentic metric, is a measure of how faithfully a dendrogram preserves the pairwise distances between the original unmodeled data points
plot(x, y, xlab = "original distance", ylab = "tree dist",
     main = "Is JC69 appropriate?", pch = 20, col = scales::alpha("black",  0.1), cex = 3)
abline(lm(y ~ x), col = "red")


x <- as.vector(mtdna.dist.n) # collapse this distance matrix
plot(x, y, xlab = "original distance", ylab = "tree dist",
     main = "Is Hammings appropriate?", pch = 20, col = scales::alpha("black",  0.1), cex = 3)
abline(lm(y ~ x), col = "red")


graphics.off()

```
Looks like Hammings or JC are appropriate. Will use JC69 based on empirical evidence/evolutionary theory.


## MLE for Final Tree
This follows the classic [Felsenstein's 1981 approach](https://link.springer.com/article/10.1007/BF01734359) as outlined by the [phangorn tutorial](https://cran.r-project.org/web/packages/phangorn/vignettes/Trees.pdf).
 
```{r}

# TODO load results

```

### Jukes-Cantor Model Fit

```{r, results='asis'}
fitJC
```

### GTR Fit
Note, I fit the process above JC69 model but am now going to maximize the Markovian DNA base-substituion model using a gamma distribution. This is the $GTR + \Gamma(4)$ (Global Time Reverse) model presented by [TavarÃ© 1986](http://www.damtp.cam.ac.uk/user/st321/CV_&_Publications_files/STpapers-pdf/T86.pdf). This model is widely considered the most flexible for a finite-sites, neutral sequence (i.e. the mtDNA).

```{r, results='asis'}
fitGTR
```

### Model Comparison
```{r, results='asis'}


out <- cbind.data.frame(model = c("JC69", "GTR"),
                        as.data.frame(anova(fitJC, fitGTR)),
                        AIC = c(AIC(fitJC), AIC(fitGTR)))

knitr::kable(out)

```

GTR model is a much better fit than JC (no surprise).


## Bootstrapping MLE-GTR Tree
Going to bootstrapped our MLE tree under the GTR model for 1e4 iterations. 
```{r}
# TODO load 
clad <- prop.clades(fitGTR$tree, bs, rooted = F)
bootval <- data.frame(node = sort(unique(fitGTR$tree$edge[,1])),
                      bs = clad)


```



## Pretty Tree
```{r}
tree <- fitGTR$tree
tidytree <-  treeio::as.treedata(tree)

#----------------------------------------------------
# Need to do some liftover work
#----------------------------------------------------
fnlsmpls <- fnlsmpls %>% 
  dplyr::rename(label = smpls)

tidytree <- full_join(tidytree, fnlsmpls, by = 'label')
tidytree@extraInfo$vividregion[is.na(tidytree@extraInfo$vividregion)] <- "internal"

# add bootstrap
bootval <- data.frame(node = sort(unique(tree$edge[,1])),
                    bs = clad)

tidytree <- full_join(tidytree, bootval, by ="node")



bstree_plotObj <- ggtree(tidytree, 
                         branch.length="none",
                         layout="circular", 
                         aes(color = vividregion)) + 
  geom_point2(aes(subset=!isTip, 
                  fill=cut(bs, c(0, 800, 900, 1000))), 
              shape=21, size=1) +
  scale_fill_manual(values=c("white", "grey", "black"), guide='legend', 
                    name='Bootstrap Percentage(BP)', 
                    breaks=c('(900,1e+03]', '(800,900]', '(0,800]'), 
                    labels=expression(BP>=90,80 <= BP * " < 90", BP < 80)) + 
  scale_color_manual("Region",
                     values = c("#fdae6b", # AF
                                "#a6bddb", #CAm
                                "#e41a1c", # DRC
                                "#c7e9c0", # EAs
                                "#000000", # internal
                                "#74c476", #ME
                                "#4a1486", # NHA
                                "#238b45", #SAm
                                "#4292c6", # SAs
                                "#084594" #SEA
                                )) + 
  theme_tree(legend.position=c(0.00, 0.5)) 

```

```{r, results='asis'}
plot(bstree_plotObj)

jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/mtdna_phylogenetic_tree.jpg", height = 8, width = 11, units = "in", res = 500)
plot(bstree_plotObj)
graphics.off()

```

