---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Section 4: Spatial Haplotype Sharing

Connections show haplotypes with one base-pair differences. The size of the circle shows the frequency of each unique haplotype within a country. 

```{r, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F, eval = T, results = 'hide',
                      fig.align = 'center', fig.width = 8, fig.height = 8)
```


```{r}
library(tidyverse)
library(sf)
library(rgeos)
library(rworldmap)
library(tidygraph)
library(ape)

```


```{r}
#---------------------------------------------------- 
# Read In Clonal Correction Files
#---------------------------------------------------- 
hapcounts <- readRDS("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/data/derived_data/mtdna_uniquehap_list.RDS") %>% 
  purrr::map(., "unihapcounts") %>% 
  dplyr::bind_rows() %>% 
  dplyr::filter(smpls != "Anc")

mtdna <- Biostrings::readDNAStringSet(filepath = "~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/data/noclonefasta/unique_mtdna.fa")
mtdna <- mtdna[ names(mtdna) != "Pcynomolgi" ]

```


```{r}
#---------------------------------------------------- 
# Read In MetaData
#---------------------------------------------------- 
smpls <- readRDS("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/data/derived_data/mtdt.RDS")

smpls <- smpls %>% 
  dplyr::filter(smpls %in% names(mtdna))


#---------------------------------------------------- 
# Read In Haplotype Counts
#---------------------------------------------------- 
smpls <- dplyr::left_join(smpls, hapcounts, by = "smpls")


```


```{r}
#---------------------------------------------------- 
# Add Jitter to Smpls
#---------------------------------------------------- 
smpls.nodrc <- smpls %>% 
  dplyr::filter(country != "CD")

smpls.nodrc <- sf::st_as_sf(smpls.nodrc, coords = c("x", "y"), crs = 4326) %>% 
  sf::st_jitter(., amount = 1.5)

# no jitter to drc, only one sample
smpls.drc <- smpls %>% 
  dplyr::filter(country == "CD") %>% 
  sf::st_as_sf(., coords = c("x", "y"), crs = 4326)


# come together
smpls <- rbind.data.frame(smpls.nodrc, smpls.drc)


```

```{r}
#---------------------------------------------------- 
# Make This a Spatial Network By Converting this to
# a SpatialLines Object
#---------------------------------------------------- 

# get combinations that will serve as endpoints for our lines
cntrylines <- smpls %>% 
  dplyr::select(c("smpls", "geometry"))
join1 <- cntrylines %>% 
  dplyr::rename(x = smpls)
join2 <- cntrylines %>% 
  dplyr::rename(y = smpls)


cntrylines.combos <- cntrylines$smpls
cntrylines.combos <- tibble::as_tibble( t( combn(cntrylines.combos, 2) ) )
cntrylines.combos <- cntrylines.combos %>% 
  magrittr::set_colnames(c("x", "y")) %>% 
  dplyr::filter( x != y)

cntrylines.combos <- dplyr::left_join(cntrylines.combos, join1, by = "x") %>% 
  dplyr::left_join(., join2, by = "y")


```

```{r}
#---------------------------------------------------- 
# Get World Background
#---------------------------------------------------- 
bb <- sf::st_bbox(
  sf::st_sf(geom = sf::st_sfc(
    sf::st_point(c(-108, -35)), # left lower
    sf::st_point(c(150, -35)), # right lower
    sf::st_point(c(-108, 50)), # left upper
    sf::st_point(c(150, 50)), # right upper
    crs = sf::st_crs("+proj=longlat +datum=WGS84 +no_defs"))
  ))


world <- rnaturalearth::ne_countries(scale = "medium", returnclass = "sf") %>% 
  sf::st_crop(., y=bb) 

```

```{r}
#---------------------------------------------------- 
# Get Hammings String Dist Between Haplotypes
#---------------------------------------------------- 
# use ape so we ignore as missing data 
mtdna.ape <- ape::as.DNAbin(mtdna)

mtdna.dist.n <- ape::dist.dna(mtdna.ape, model="N")


mtdnadistlong <- broom::tidy(mtdna.dist.n) %>% 
  magrittr::set_colnames(c("x", "y", "hammings"))

# get complement comparisons for easier join
revmtdnadistlong <- mtdnadistlong %>% 
  dplyr::select(c("y", "x", "hammings")) %>% 
  magrittr::set_colnames(c("x", "y", "hammings"))

mtdnadistlong <- rbind.data.frame(mtdnadistlong, revmtdnadistlong) 

# add in some details for sanity
join1 <- smpls %>% 
  dplyr::rename(x = smpls) %>% 
  dplyr::select(c("x", "country"))
sf::st_geometry(join1) <- NULL
join2 <- smpls %>% 
  dplyr::rename(y = smpls) %>% 
  dplyr::select(c("y", "country"))
sf::st_geometry(join2) <- NULL
mtdnadistlong <- mtdnadistlong %>% 
  dplyr::left_join(., join1, by = "x") %>% 
  dplyr::left_join(., join2, by = "y")

```

### All Hamming's Distances
```{r}
#---------------------------------------------------- 
# Bind Distance and Space
#---------------------------------------------------- 
drcsmpls <- smpls$smpls[smpls$country == "CD"]
nhasmpls <- smpls$smpls[smpls$vividregion == "NHA"]

mtdnanetwork <- dplyr::left_join(cntrylines.combos, mtdnadistlong, by = c("x", "y")) 

mtdnanetwork.diff <- mtdnanetwork %>% 
  dplyr::filter(x %in% drcsmpls | y %in% drcsmpls) %>% # only look at DRC
 # dplyr::filter(hammings ==1) %>%  # have to be within 1 bp  
  dplyr::mutate(long1 = sf::st_coordinates(geometry.x)[,1],
                lat1 =  sf::st_coordinates(geometry.x)[,2], 
                long2 = sf::st_coordinates(geometry.y)[,1],
                lat2 =  sf::st_coordinates(geometry.y)[,2])



points <- smpls %>% 
  dplyr::mutate(country = ifelse(vividregion == "NHA", "NHA", country),
                x = sf::st_coordinates(.)[,1],
                y = sf::st_coordinates(.)[,2])
  
# seperate out drc and other for plotting
points.nodrc <- points %>% 
  dplyr::filter(country != "CD")
points.drc <- points %>% 
  dplyr::filter(country == "CD")

# cool to warm colors
# cool <- rainbow(50, start=rgb2hsv(col2rgb('cyan'))[1], end=rgb2hsv(col2rgb('blue'))[1])
# warm <- rainbow(50, start=rgb2hsv(col2rgb('red'))[1], end=rgb2hsv(col2rgb('yellow'))[1])
# cols <- c(rev(cool), rev(warm))
# pal <- colorRampPalette(cols)(6)

mtDNAglobal <- ggplot() +
  geom_sf(data = world, color = "#ffffff", fill = "#000000", size = 0.05) +
  geom_curve(data = mtdnanetwork.diff, alpha = 0.5, size = 1.1,
             aes(x = long1, y = lat1, xend = long2, yend = lat2, color = factor(hammings))) +
  geom_point(data = points.nodrc, aes(x = x, y = y, 
                                      size = hapuid_count), shape = 21, stroke = 0, show.legend = F, alpha = 0.8, fill = "#d9d9d9") +
  geom_point(data = points.drc, aes(x = x, y = y), size = 3, shape = 17, color = "#f0f0f0",  show.legend = F) +
  scale_color_viridis_d("Hamming's Distance", option="plasma", begin = 1, end = 0) + 
  scale_size_continuous(range = c(1.4,4)) +
  theme(plot.title = element_blank(),
        axis.title = element_blank(),
        axis.text = element_blank(),
        legend.position = "bottom",
        legend.title = element_text(family = "Helvetica", hjust = 0.5, vjust = 0.5, size = 12),
        legend.text = element_text(family = "Helvetica", hjust = 0.5, vjust = 0.5, size = 10),
        plot.background = element_rect(fill = "transparent"),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "#deebf7")) +
  coord_sf(xlim = c(bb['xmin'], bb['xmax']), 
           ylim = c(bb['ymin'], bb['ymax']), 
           datum = NA) +
  ggspatial::annotation_north_arrow(location = "bl", which_north = "true", size = 0.8) +
  guides(color=guide_legend(nrow=1, byrow = T))

```


```{r, results='asis', fig.align='center', fig.width = 11, fig.height = 8}
plot(mtDNAglobal)

jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/mtdna_spatial_network_wide.jpg", height = 8, width = 11, units = "in", res = 500)
plot(mtDNAglobal)
graphics.off()

jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/mtdna_spatial_network_long.jpg", height = 11, width = 8, units = "in", res = 500)
plot(mtDNAglobal)
graphics.off()


```



### One-Hammings Distance
```{r}
#---------------------------------------------------- 
# Bind Distance and Space
#---------------------------------------------------- 

mtdnanetwork.diff <- mtdnanetwork %>% 
  dplyr::filter(x %in% drcsmpls | y %in% drcsmpls) %>% # only look at DRC
  dplyr::filter(hammings ==1) %>%  # have to be within 1 bp  
  dplyr::mutate(long1 = sf::st_coordinates(geometry.x)[,1],
                lat1 =  sf::st_coordinates(geometry.x)[,2], 
                long2 = sf::st_coordinates(geometry.y)[,1],
                lat2 =  sf::st_coordinates(geometry.y)[,2])




mtDNAglobal <- ggplot() +
  geom_sf(data = world, color = "#ffffff", fill = "#000000", size = 0.05) +
  geom_curve(data = mtdnanetwork.diff, alpha = 0.5, size = 1.1,
             aes(x = long1, y = lat1, xend = long2, yend = lat2), color = "#FF0000", show.legend = F) +
  geom_point(data = points.nodrc, aes(x = x, y = y, 
                                      size = hapuid_count), shape = 21, stroke = 0, show.legend = F, alpha = 0.8, fill = "#d9d9d9") +
  geom_point(data = points.drc, aes(x = x, y = y), size = 3, shape = 17, color = "#f0f0f0",  show.legend = F) +
  scale_size_continuous(range = c(1.4,4)) +
  theme(plot.title = element_blank(),
        axis.title = element_blank(),
        axis.text = element_blank(),
        legend.position = "bottom",
        legend.title = element_text(family = "Helvetica", hjust = 0.5, vjust = 0.5, size = 12),
        legend.text = element_text(family = "Helvetica", hjust = 0.5, vjust = 0.5, size = 10),
        plot.background = element_rect(fill = "transparent"),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "#deebf7")) +
  coord_sf(xlim = c(bb['xmin'], bb['xmax']), 
           ylim = c(bb['ymin'], bb['ymax']), 
           datum = NA) +
  ggspatial::annotation_north_arrow(location = "bl", which_north = "true", size = 0.8) 

```


```{r, results='asis', fig.align='center', fig.width = 11, fig.height = 8}
plot(mtDNAglobal)

jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/mtdna_spatial_network_wide_hammingone.jpg", height = 8, width = 11, units = "in", res = 500)
plot(mtDNAglobal)
graphics.off()

jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/mtdna_spatial_network_long_hammingone.jpg", height = 11, width = 8, units = "in", res = 500)
plot(mtDNAglobal)
graphics.off()


```



#### Brazil Comparison
```{r}
#---------------------------------------------------- 
# Run Brazil Comparison
#---------------------------------------------------- 
BRsmpls <- smpls$smpls[smpls$country == "BR"]

mtdnanetwork.diff <- mtdnanetwork %>% 
  dplyr::filter(x %in% BRsmpls | y %in% BRsmpls) %>% # only look at DRC
  # dplyr::filter(hammings ==1) %>%  # have to be within 1 bp  
  dplyr::mutate(long1 = sf::st_coordinates(geometry.x)[,1],
                lat1 =  sf::st_coordinates(geometry.x)[,2], 
                long2 = sf::st_coordinates(geometry.y)[,1],
                lat2 =  sf::st_coordinates(geometry.y)[,2])


# seperate out drc and other for plotting
points.nobr <- points %>% 
  dplyr::filter(country != "BR")
points.br <- points %>% 
  dplyr::filter(country == "BR")


BRmtDNAglobal <- ggplot() +
  geom_sf(data = world, color = "#ffffff", fill = "#000000", size = 0.05) +
  geom_curve(data = mtdnanetwork.diff, alpha = 0.5, size = 1.1,
             aes(x = long1, y = lat1, xend = long2, yend = lat2, color = factor(hammings))) +
  geom_point(data = points.nobr, aes(x = x, y = y, 
                                      size = hapuid_count), shape = 21, stroke = 0, show.legend = F, alpha = 0.8, fill = "#d9d9d9") +
  geom_point(data = points.br, aes(x = x, y = y), size = 3, shape = 17, color = "#f0f0f0",  show.legend = F) +
  scale_color_viridis_d("Hamming's Distance", option="plasma", begin = 1, end = 0) + 
  scale_size_continuous(range = c(1.4,4)) +
  theme(plot.title = element_blank(),
        axis.title = element_blank(),
        axis.text = element_blank(),
        legend.position = "bottom",
        legend.title = element_text(family = "Helvetica", hjust = 0.5, vjust = 0.5, size = 12),
        legend.text = element_text(family = "Helvetica", hjust = 0.5, vjust = 0.5, size = 10),
        plot.background = element_rect(fill = "transparent"),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "#deebf7")) +
  coord_sf(xlim = c(bb['xmin'], bb['xmax']), 
           ylim = c(bb['ymin'], bb['ymax']), 
           datum = NA) +
  ggspatial::annotation_north_arrow(location = "bl", which_north = "true", size = 0.8) +
  guides(color=guide_legend(nrow=1, byrow = T))

```


```{r, results='asis', fig.align='center', fig.width = 11, fig.height = 8}
plot(BRmtDNAglobal)

jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/mtdna_spatial_network_wide_BRAZIL.jpg", height = 11, width = 8, units = "in", res = 500)
plot(BRmtDNAglobal)
graphics.off()


```

#### Malaysia Comparison
```{r}
#---------------------------------------------------- 
# Run Malaysia Comparison
#---------------------------------------------------- 
MYsmpls <- smpls$smpls[smpls$country == "MY"]

mtdnanetwork.diff <- mtdnanetwork %>% 
  dplyr::filter(x %in% MYsmpls | y %in% MYsmpls) %>% 
  # dplyr::filter(hammings ==1) %>%  # have to be within 1 bp  
  dplyr::mutate(long1 = sf::st_coordinates(geometry.x)[,1],
                lat1 =  sf::st_coordinates(geometry.x)[,2], 
                long2 = sf::st_coordinates(geometry.y)[,1],
                lat2 =  sf::st_coordinates(geometry.y)[,2])


# seperate out drc and other for plotting
points.nomy <- points %>% 
  dplyr::filter(country != "MY")
points.my <- points %>% 
  dplyr::filter(country == "MY")


MYmtDNAglobal <- ggplot() +
  geom_sf(data = world, color = "#ffffff", fill = "#000000", size = 0.05) +
  geom_curve(data = mtdnanetwork.diff, alpha = 0.5, size = 1.1,
             aes(x = long1, y = lat1, xend = long2, yend = lat2, color = factor(hammings))) +
  geom_point(data = points.nomy, aes(x = x, y = y, 
                                      size = hapuid_count), shape = 21, stroke = 0, show.legend = F, alpha = 0.8, fill = "#d9d9d9") +
  geom_point(data = points.my, aes(x = x, y = y), size = 3, shape = 17, color = "#f0f0f0",  show.legend = F) +
  scale_color_viridis_d("Hamming's Distance", option="plasma", begin = 1, end = 0) + 
  scale_size_continuous(range = c(1.4,4)) +
  theme(plot.title = element_blank(),
        axis.title = element_blank(),
        axis.text = element_blank(),
        legend.position = "bottom",
        legend.title = element_text(family = "Helvetica", hjust = 0.5, vjust = 0.5, size = 12),
        legend.text = element_text(family = "Helvetica", hjust = 0.5, vjust = 0.5, size = 10),
        plot.background = element_rect(fill = "transparent"),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        panel.background = element_rect(fill = "#deebf7")) +
  coord_sf(xlim = c(bb['xmin'], bb['xmax']), 
           ylim = c(bb['ymin'], bb['ymax']), 
           datum = NA) +
  ggspatial::annotation_north_arrow(location = "bl", which_north = "true", size = 0.8) +
  guides(color=guide_legend(nrow=1, byrow = T))

```


```{r, results='asis', fig.align='center', fig.width = 11, fig.height = 8}
plot(MYmtDNAglobal)

jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/mtdna_spatial_network_wide_MALAYSIA.jpg", height = 11, width = 8, units = "in", res = 500)
plot(MYmtDNAglobal)
graphics.off()


```

```{r}

BRmtDNAglobal <- BRmtDNAglobal + theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))
MYmtDNAglobal <- MYmtDNAglobal + theme(plot.margin = unit(c(0, 0, 0, 0), "cm"))

jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/mtdna_spatial_network_wide_comparisoins.jpg", height = 11, width = 8, units = "in", res = 500)
cowplot::plot_grid(BRmtDNAglobal, MYmtDNAglobal, align = "h", ncol = 1)
graphics.off()
```

