---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Erbo Spain, 1944 Isolate
```{r}
library(tidyverse)
library(Biostrings)
library(vcfR)
source("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/R/vcf_utils.R")
```

```{r}
#........................................................
# Read in fastas
#........................................................
pvp01 <- Biostrings::readDNAStringSet("~/Documents/MountPoints/mountIDEEL/resources/genomes/Pvivax/genomes/PvP01.fasta")
                            
pvp01 <- pvp01[ names(pvp01) == "PvP01_MIT_v1" ]

```
```{r}
#.................................
# Read in data
#.................................
ebro <- vcfR::read.vcfR("~/Documents/MountPoints/mountedScratchLL/Projects/VivID_Seq/Erbo1944/Erbo1944.vcf")
pos <- readRDS("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/data/derived_data/final_positions.rds")

#.................................
# Read in Metadata
#.................................
mtdt <- readRDS("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/data/derived_data/mtdt.RDS")


ebro <- data.frame(iid = "Ebro-1944", smpls = "Ebro-1944", country = "ES", vividregion = "Ebro-1944", passed = "Y", hapuid = NA, x = NA, y = NA, k = NA)

mtdt <- rbind.data.frame(mtdt, ebro)

```

```{r}
#........................................................
# SNPs only
#........................................................
erbo.snp <- vcfR::extract.indels(erbo, return.indels = F)
erbo.snp

#........................................................
# Get POS
#........................................................
erboliftover <- data.frame(POS = vcfR::getPOS(erbo.snp),
                           GT = vcfR::extract.gt(erbo.snp),
                           REF = vcfR::getREF(erbo.snp),
                           ALT = vcfR::getALT(erbo.snp))
erboliftover <- erboliftover %>% 
  dplyr::filter(POS %in% pos)

#........................................................
# Do liftover
#........................................................
erbo <- pvp01
for(i in 1:length(erboliftover$POS)){
  erbo[[1]][erboliftover$POS[i]] <- as.character( erboliftover$ALT[i] ) # all 1/1
}
```


```{r}

mtdna <- Biostrings::readDNAStringSet(filepath = "~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/data/fasta/mtdna_anc.fa")

# going to drop the ancestral haplo here because I will do an unrooted tree
# since we are not estimating the molecular clock
mtdna <- mtdna[!names(mtdna) %in% "Pcynomolgi"]

# going to add erbo
mtdna[length(mtdna) + 1] <- erbo
names(mtdna)[length(mtdna)] <- "Ebro-1944"

seqmat <-  tibble::as_tibble( t(as.matrix(mtdna) ) )
seqmat <- cbind.data.frame(POS = 1:nrow(seqmat), seqmat)

segsites <- apply(seqmat[,2:ncol(seqmat)], 1, function(x){ return( length(unique(x)) > 1 ) })


seqmat.long <- seqmat %>% 
  dplyr::filter(segsites) %>% 
  tidyr::gather(., key = "smpls", value = "GT", 2:ncol(.)) %>% 
  dplyr::mutate(GT = factor(GT)) %>% 
  dplyr::left_join(., mtdt, by = "smpls") %>% 
  dplyr::filter(passed == "Y") %>% 
  dplyr::mutate(POS = factor(POS, levels = c(1:5989), ordered = T)) 

gtplot <- seqmat.long %>% 
  ggplot() + 
  geom_tile(aes(x = smpls, y = POS, fill = GT)) + 
  scale_fill_manual("Alleles", values = c("#4daf4a", "#377eb8", "#ffff33", "#e41a1c")) +
  facet_wrap(~vividregion + country, scales = "free_x", nrow = 1, strip.position = "top") +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank())

jpeg("~/Documents/GitHub/VivID_Seq/PopGenome_Analysis/figures/ebro_compare.jpg", height = 8, width = 11, units = "in", res = 500)
plot(gtplot)
graphics.off()

```

## Spain 1944 Visualization
Isolate from near the Ebro River in Spain dating to 1944 (microscopy slide collection from the Victorian Era).
```{r, results='asis', fig.width=11, fig.height=8, fig.align='center'}
plot(gtplot)
```







